shard/connpool.lua

```diff
@@ -20,45 +20,99 @@ function heartbeat(pool_id)
 end
 
 -- default callbacks
-local function on_connfail(self, srv)
-    log.info('%s connection failed', srv.uri)
+
+-- called when couldn't retrieve heartbeat
+local function on_monitor_failure(self, srv)
+    log.warn('- %s - connection failed', srv.uri)
+end
+
+-- called during first connection in case of an error
+local function on_connection_failure(self, srv)
+    srv.state = 'failed'
+    log.warn('- %s - connection check failed', srv.uri)
 end
 
+-- on server connect
 local function on_connected_one(self, srv)
+    srv.state = 'connected'
     log.info(' - %s - connected', srv.uri)
 end
 
+-- called from wait_connection when all servers are connected
 local function on_connected(self)
     log.info('connected to all servers')
 end
 
+-- on server disconnect
 local function on_disconnect_one(self, srv)
-    log.info("kill %s by dead timeout", srv.uri)
+    log.warn("kill %s by dead timeout", srv.uri)
 end
 
+-- called when the whole zone is disconnected
 local function on_disconnect_zone(self, name)
-    log.info("zone %s has no active connections", name)
+    log.warn("zone %s has no active connections", name)
+    log.error("Cluster is invalid. You must turn on read only mode")
 end
 
+-- called when each node in connection pool is disconnected
 local function on_disconnect(self)
-    log.info("there is no active connections")
+    log.error("there is no active connections")
+end
+
+-- actions to check for connpool's state on server disconnect
+local function _on_disconnect(self, srv)
+    self:on_disconnect_one(srv)
+
+    -- check zone and pool
+    local d = 0
+    local all_zones = self:zone_list()
+    for _, name in pairs(all_zones) do
+        local alive = self:get_all_active_servers(name)
+        if #alive == 0 then
+            self:on_disconnect_zone(name)
+            d = d + 1
+        end
+    end
+    if d == #all_zones then
+        self:on_disconnect()
+    end
 end
 
+-- called when init is finished
+-- connections may not be established at this point
 local function on_init(self)
+    self.init_complete = true
     log.info('started')
 end
 
+-- called when node is acknowleged dead by all other nodes in pool
+local function on_dead_disconnected(self, srv)
+    log.warn("Server %s is marked dead", srv.uri)
+    srv.state = 'is_dead_disconnected'
+    if on_server_fail then
+        log.warn("Initializing failover procedure")
+        on_server_fail(srv)
+    end
+    self.epoch_counter = self.epoch_counter + 1
+    _on_disconnect(self, srv)
+end
 
-local function server_is_ok(self, srv, dead)
-    if srv.ignore then
-        return false
+-- called when once proven dead node returns online
+local function on_dead_connected(self, srv)
+    srv.state = 'is_dead_connected'
+    if on_server_return then
```

Comment:
What is on_server_return? I do not see these name in connpool.lua.

Reply:
it's a method one has to implement for custom failover procedures